contracts/BluzelleESR.sol

```diff
@@ -128,12 +128,38 @@ contract BluzelleESR {
 
     //returns number of nodes in a swarm
     function getNodeCount(string memory swarmID) public view onlyIfActive() returns(uint256) {
-        return (SwarmStructs[swarmID].nodeList.length);
+
+        uint i;
+        uint256 trueCount = 0;
+        bytes memory nodeEntryBytes;
+        
+        for(i=0; i<SwarmStructs[swarmID].nodeList.length; i++){
+            nodeEntryBytes = bytes(SwarmStructs[swarmID].nodeList[i]);
+            if(nodeEntryBytes.length != 0)
```

Comment:
How do these empty values happen in the first place? Is it possible to prevent them from being added instead of doing this?

Reply:
When you want to modify the size of the array (like adding an element), you have to pay ether (anytime you modify the contract on the blockchain actually).  There is a method to delete the elements (the data) in the array but it won't delete the memory allocation on the blockchain, thus leaving that empty "gap" in the array.

We COULD reassign the array to a new array whenever an element is deleted (excluding the gaps) OR shift the elements to take out the gaps.  Everytime you do that, you will be paying ether for it.

We could do this.  The option that I was thinking was no cost for us.  getNodeCount and getSwarmList will just ignore the gaps.  getNodeList and getSwarmList function will still show the gaps in the respective arrays, but it can easily be handled on client side, which is free.

Reply:
We might also take it a step further and initialize the array with some fixed upper-bound, say 100 nodes. Then, if I understand correctly, we avoid paying ether for both creates and deletes.

Reply:
Sorry I might have been a bit unclear with my previous comment.

Any time you write to the contract (modifying the contract), you have to pay ether.  So when you add data (writing and storing "hello world" to the contract on the blockchain) (even if the array has a fixed upper-bound), you'd still need to pay ether.

When you delete data from the array, in solidity, the "delete" reserved keyword pretty much just takes the element data bits and replaces it with a 0.  So for example, if the first element of the array is "hello world" which is 0x123 bits, applying delete on the first element will convert 0x123 to 0.  This will result with a "gap" where the first element should be in the array.  This operation is free (I think it was explained as "resetting" the bits).

If we wanted to explicitly remove that first element gap, then we are modifying the array, which means we are modifying the contract, which means paying ether.